### DISCUSSION TITLE
Custom proxy handler

### DISCUSSION THREAD

**Discussion by jacob-orbiit** [pmndrs/valtio#610] [Ideas]
I may be creating an [X-Y problem](https://en.wikipedia.org/wiki/XY_problem) for myself, but the scenario I'm looking for is "use data if it exists, otherwise fetch it" (with a manual override, like `refresh()`). This works well for fields with known names like `user`:

```js
let gettingMe;

const state = proxy({ /* … */ });

const store = derive({
  async user(get) {
    const { user } = get(state);

    if (user.isAuthenticated) return user;

    gettingUser ??= getUser();

    await gettingUser;

    return user;
  },
});
```

However, it's quite common to have collections keyed by ids. If I were setting up the `Proxy` myself for that, I would do something like:

```ts
let gettingProduct;

new Proxy(new Map<ProductUUID, Product>(), {
  async get(products, prop) {
    if (products.has(prop)) return products.get(prop);

    if (!gettingProduct.has(prop)) getProduct(prop);

    await gettingProduct;

    return products.get(prop);
  },
});
```

I read through the docs and README, but didn't find anything that seemed adequate for facilitating this; my apologies if this already exists.

**Comment by dai-shi** [comment]
Let's see if you really need proxies for this case.

This works capability-wise, right?
```js
const state = proxy({ user: null })

const getUserFromState = async () => {
  if (!state.user) {
    state.user = await getUser()
  }
  return state.user
}
```

**Reply by jacob-orbiit** [reply]
Yes, that would certainly work, but I'm hoping for something transparent to the consumer. The beauty of valtio (as I see it is) it enables the consumer to be blissfully ignorant of the pipes and consume state as a normal object: You want user? `state.user`. You want to update user? `state.user = newUser` (under the hood, `derive({ get user() {…}, set user() {…} })`). Behind the scenes, the store transparently handles any persistence (WebStorage, remote API, etc).

**Reply by dai-shi** [reply]
We have a similar discussion lately: https://github.com/pmndrs/valtio/discussions/609#discussioncomment-4463224

While I understand your desire, you seem to have a misconception. Valtio proxy doesn't do anything for `get` handler. It only traps `set` handler. This is a fundamental design decision to make the core library small, predictable and maintainable.

So, what's requested is essentially out of the scope, but we give some extensibility so that users and third-party libraries can add capabilities in the ecosystem.

In this case, some options are:
1. `proxyWithComputed` with `proxy-memoize`
2. Wrapping a valtio proxy with your own proxy (what I call double proxies)
3. `unstable_buildProxyFunction`

**Reply by jacob-orbiit** [reply]
Hmm, I'm trying the double-proxy option, like so:

```js
const state = proxy({
  // …
  products: new Proxy(proxyMap(), {
    get(products, prop, receiver) {
      if (typeof prop === 'symbol') return products[prop];

      switch (typeof products[prop]) {
        case 'function': return Reflect.apply(products[prop], receiver, []);
        case 'number': return Reflect.get(products, prop, receiver);
      }

      const product = Reflect.get(products, prop, receiver);

      if (product?.isFullyLoaded) { return product }

      if (!gettingProduct.has(prop)) getProduct(prop);

      return gettingProduct
        .get(prop)
        .then(() => products.get(prop));
    },
  }),
});

// const store same as above: derive({…})
```

but `prop` is often `'data'`, which I don't recognise—my app is not trying to access `products.data` and AFAIK, `'data'` is not a standard prop on `Map()`.

(`prop` is also sometimes `'then'`, I think because of the promise returned from `gettingProduct`, which I haven't figured out yet)

**Reply by dai-shi** [reply]
I think it's easier to have your proxy outside the valtio proxy.
```js
new Proxy(proxy(...), { get() { ... } })
```

I'd also suggest to avoid `proxyMap` in this case which is just an emulation around normal object/array.

**Reply by jacob-orbiit** [reply]
Hmm, that is giving me very bizarre results: `prop` is always a symbol (I'm guessing it's from an interator?) and `Reflect.get(target, prop, receiver)` appears to be returning internals of valtio (specifically, [`proxyState`](https://github.com/pmndrs/valtio/blob/main/src/vanilla.ts#L273-L278)).

```js
export const state = proxy({
  errors: new Array(0),
  user: {
    id: '',
    isAuthenticated: false,
    // …
  },
});
export const store = new Proxy(state, {
  get(target, prop, receiver) {
    const result = Reflect.get(target, prop, receiver);

    // [
    //   { errors: Proxy, user: Proxy },
    //   Proxy { errors: Proxy, user: Proxy },
    //   20,
    //   (target, receiver, version2, handlePromise = defaultHandlePromise) => {…},
    //   4,
    //   Set(0) { size: 0 },
    // ]

    return result;
  },
  set(target, prop, value) {
    const result = Reflect.set(target, prop, value);

    return result;
  },
});
```

Whatever that bizarre-looking output is, valtio _seems_ to expect it as `user` from `const { user } = useSnapshot(store)` is the initial value in the above `state`. However, the traps above are never called with any useful values to catch (ex I would expect for `prop` to be `'user'` at some point, as it was in the previous iteration with the `new Proxy()` inside valtio's `proxy()`).

Perhaps it would be better to make smaller states/stores? (I'm trying that out now)

```js
export const errorsState = proxySet();
export const productsState = new Proxy(proxy(), {
  get() { },
  has() { },
  set() { },
});
export const userState = proxy();
```

P.S. Thanks for helping out with this! I know you said it's explicitly out of scope for valtio, but if I can get this working, I think it would be valuable to document in the repo's wiki (I'd be happy to contribute that).

**Reply by jacob-orbiit** [reply]
So it seems these are all "six of one, half a dozen of the other", wherein the `Proxy()`'s `get` trap is only ever invoked at the top-level of getting the collection itself (not items on the collection).

**Reply by jacob-orbiit** [reply]
Ah, I think I've got it:

```js
const promiseMethodNames = new Set(['catch', 'finally', 'then']);
const products = new Proxy(proxy({}), {
  get(collection, prop, receiver) {
    // Valtio `await`s the parent promise, and await continuously calls `then()`
    // until it reaches the end of the chain (wherein `then()` is undefined).
    if (promiseMethodNames.has(prop)) { return }

    const field = Reflect.get(collection, prop, receiver);
    if (typeof prop === 'symbol' || field) { return field }

    if (prop.length === 24) { return getProduct(prop) }
  },
  set(collection, id, value) {…},
});
export const state = proxy({
  errors,
  products: new Promise((resolve) => {
    if (Object.keys(products).length) return resolve(products);

    gettingProducts ??= getProducts();

    return resolve(gettingProducts.then(() => products));
  }),
  user: new Promise((resolve) => {
    if (user.isAuthenticated) return resolve(user);

    gettingUser ??= getUser();

    return resolve(gettingUser.then(() => user));
  }),
});
```

**Reply by dai-shi** [reply]
Hm, I didn't expect so complicated. Maybe because it uses promises.

**Comment by dai-shi** [comment]
https://github.com/pmndrs/valtio/discussions/610#discussioncomment-4463002

Let's change it to non async, and store a promise.

```js
const state = proxy({ user: null })

const getUserFromState = () => {
  if (!state.user) {
    state.user = getUser()
  }
  return state.user
}
```

If that ☝️ works, the proxy version would be something like this 👇 .
```js
const state = new Proxy(proxy({ user: null }), {
  get(target, prop) {
    if (prop === 'user' && !target.user) {
      target.user = getUser()
    }
    return target[prop]
  },
})
```

**Reply by jacob-orbiit** [reply]
the `user` part of state was not the issue—it's items on a collection where the prop name is unknown:

```ts
interface State {
  products: Record<string, Product>;
  user: User;
}
```

`state.products` would ideally be `Map<string, Product>`, but that hugely complicated the `Proxy()`.

I tried removing the `new Promise()` from the parent `state`, and that caused an infinite loop:

```js
const products = new Proxy(proxy({}), { get(…) {…}, set(…) {…});
const user = proxy({ id: '', isAuthenticated: false });

export const state = proxy({
  errors,
  get products() {
    if (Object.keys(products).length) return products;

    gettingProductsList ??= getProductsList();

    return gettingProductsList.then(() => products);
  },
  get user() {
    if (user.isAuthenticated) return user;

    gettingMe ??= getMe();

    return gettingMe.then(() => user);
  },
});
```

**Reply by dai-shi** [reply]
Do you want to fetch product each on demand, or can you fetch all products at once?

**Reply by jacob-orbiit** [reply]
Fetch on demand

**Reply by dai-shi** [reply]
Something like this?
```js
const getProducts = () => new Proxy(proxy({}), {
  get(target, prop) {
    if (!target[prop]) {
      target[prop] = getProduct(prop)
    }
    return target[prop]
  },
})

const state = new Proxy(proxy({ user: null, products: null }), {
  get(target, prop) {
    if (prop === 'user' && !target[prop]) {
      target[prop] = getUser()
    }
    if (prop === 'products' && !target[prop]) {
      target[prop] = getProducts()
    }
    return target[prop]
  },
})
```

**Reply by jacob-orbiit** [reply]
I think that's basically the same as what I posted above (https://github.com/pmndrs/valtio/discussions/610#discussioncomment-4509243), except this has an extra `Proxy()` wrapper (on `state`, which I think is not necessary?) and consolidates each top-level field into 1 big getter instead of each their own.

P.S. Sorry for the delay—I was travelling.

**Reply by dai-shi** [reply]
okay, so basically we reached the same line.
